### DISCUSSION TITLE
Multiple projects causing AuthStore conflicts.

### DISCUSSION THREAD

**Discussion by StickOnAStick** [pocketbase/pocketbase#2147] [Q&A] [answered]
I'm currently working on two nextJS13 projects, with both using their own pocketbase executable. Project A was created before Project B, both have their own separate adminUi, collections, and records. Both adminUi were created and contain a user with the same email.  

When working on Project B _(with Project A shutdown)_ and trying to call ```authStore.model.id``` I receive the userID associated with the user in Project A. This makes user lookup via authstore provided id impossible. 

**Fixes tried:**

- Removing pb_data from Project A
- Removing the backend folder entirely from Project A
- Clearing authstore before fetching model. 

AuthStore.model inside Proj. B:

![Untitled](https://user-images.githubusercontent.com/49598942/227476726-3c809090-b529-49ab-a80a-7dd6b8b4aec8.png)

#2145

**Comment by ganigeorgiev** [comment, marked as answer]
The `localStorage` state is shared for the same `protocol://host:port`.

I assume that you are testing on localhost and that would explain the same `pb.authStore.model` for your 2 projects.

I don't understand how you manage your `pb_data` and how it is related to your 2 different projects.

By default the executable will use the `pb_data` located next to it. If you want to specify an explicit `pb_data` location you can use the `--dir` flag (_for list of all flags you can run `./pocketbase serve --help`_).

**Reply by ganeshkbhat** [reply]
I probably that this not completely right. the `nid` rfc structure is similar but the localStorge is stored as for a path as `[protocol]://[subdomain].[[domain].[extension]]:[port]/path` only for https and `https://www.university.com/path` as following but is absent for udp:

![https-udp-data-format](https://user-images.githubusercontent.com/3424344/227487504-d7b79b94-399e-4270-82af-8e3709813252.png)

**Reply by ganigeorgiev** [reply]
@ganeshkbhat I have no idea what do you mean and how it is related to the question. If you mean that in the above `protocol` should be `scheme`, then yes, that probably would be more correct.

`localStorage` is scoped per origin. From the [localStorage MDN argicle](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage):

> A [Storage](https://developer.mozilla.org/en-US/docs/Web/API/Storage) object which can be used to access the current origin's local storage space.
> ...
> localStorage data is specific to the protocol of the document.

The definition for origin per the SameOriginPolicy is:

> An origin is defined by the scheme, host, and port of a URL

**Reply by ganeshkbhat** [reply]
@ganigeorgiev you are correct. please do not put `protocol://host:port` it was a bit misleading. please put urlpath as `[httpprotocol]://[domainhost].[ext]:[domainport]/path` or `[httpprotocol]://[subdomainhost].[domainhost].[ext]:[domainport]/path` as in [https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin). The `protocol://host:port` does not apply to other protocols for localstorage and may be misleading/absent for other protocols.

**Reply by ganigeorgiev** [reply]
The terms "protocol" and "scheme" overlap and there is no need to nitpick. Even from your linked resource:

> Two URLs have the same origin if the [protocol](https://developer.mozilla.org/en-US/docs/Glossary/Protocol), [port](https://developer.mozilla.org/en-US/docs/Glossary/Port) (if specified), and [host](https://developer.mozilla.org/en-US/docs/Glossary/Host) are the same for both.

In any case, the idea is that if both Next.js projects are using `http://127.0.0.1:3000` they will share the same `localStorage`.

**Reply by StickOnAStick** [reply]
> The `localStorage` state is shared for the same `protocol://host:port`.
> 
> I assume that you are testing on localhost and that would explain the same `pb.authStore.model` for your 2 projects.
> 
> I don't understand how you manage your `pb_data` and how it is related to your 2 different projects.
> 
> By default the executable will use the `pb_data` located next to it. If you want to specify an explicit `pb_data` location you can use the `--dir` flag (_for list of all flags you can run `./pocketbase serve --help`_).

### Dev environments for both projects: 

```
Proj A & B: (seperate directories)
     | - - Backed:
     |      | - - pb_data
     |      | - - ...
     |      | - - pocketbase.exe
     |
     | - - Frontend
```

Project B's pb_data directory is manually set using `./pocketbase serve -dir .../ProjB/Backend/pb_data`. As you'd mentioned, I am using localhost for both projects but had assumed that the two projects would still have separate authstores due to the separate data directories. 

# Fix
Since you'd mentioned that `localstorage` is shared for the same port, I'd switch Proj B's PB port using `./pocketbase serve --http "127.0.0.1:8091"` and things are working perfectly. 

Thank you for your information and help with this issue! @ganigeorgiev

**Comment by ganeshkbhat** [comment]
> The terms "protocol" and "scheme" overlap and there is no need to nitpick. Even from your linked resource:
> 
> > Two URLs have the same origin if the [protocol](https://developer.mozilla.org/en-US/docs/Glossary/Protocol), [port](https://developer.mozilla.org/en-US/docs/Glossary/Port) (if specified), and [host](https://developer.mozilla.org/en-US/docs/Glossary/Host) are the same for both.
> 
> In any case, the idea is that if both Next.js projects are using `http://127.0.0.1:3000` they will share the same `localStorage`.

not nitpicking, just sharing. I love your project. It would have taken me miles to even do a kudos. i would love to join you as a contributor/ maintainer if you are looking for one long term. i am intermediate plus with python and typescript and just started with go lang. probably not the right place but could not find any better place.
